rbac/management/authorization/token_validator.py

```diff
@@ -0,0 +1,147 @@
+# The service accounts claim we are expecting to find in the token.
+SERVICE_ACCOUNTS_CLAIM = "api.iam.service_accounts"
+
+
+class ITSSOTokenValidator:
+    """JWT token  validator."""
+
+    def __init__(self):
+        """Get the OIDC configuration URL."""
+        # TODO replace it with:
+        it_host = settings.IT_SERVICE_HOST
+        it_port = settings.IT_SERVICE_PORT
+        it_scheme = settings.IT_SERVICE_PROTOCOL_SCHEME
+
+        # it_host = "https://sso.stage.redhat.com"
+
+        # The host contains the URL including the port...
+        self.host = f"{it_scheme}://{it_host}:{it_port}/auth/realms/redhat-external"
+        # ... but the issuer does not. We need to make this distinction so that when validating the token, the issuer
+        # correctly matches.
+        self.issuer = f"{it_scheme}://{it_host}/auth/realms/redhat-external"
+        self.oidc_configuration_url = f"{self.host}/.well-known/openid-configuration"
+
+    def validate_token(self, request: Request) -> str:
+        """Validate the JWT token issued by Red Hat's SSO.
+
+        Performs validations on the issuer, audience and scope of the token. Raises exceptions if the token is not
+        valid. Finally, it returns the received Bearer token in order to be able to forward it as is to IT.
+        """
+        if settings.IT_BYPASS_TOKEN_VALIDATION:
+            return "mocked-invalid-bearer-token-because-token-validation-is-disabled"
+
+        bearer_token: str = request.headers.get("Authorization")
+        if not bearer_token:
+            logger.debug(
+                f"Issuing unauthorized response either because the Authorization header is missing, or it's blank."
+                f" Request: {request}"
+            )
+            raise MissingAuthorizationError()
+
+        # Strip the "Bearer" part of the token if it comes with it.
+        if bearer_token.startswith("Bearer"):
+            bearer_token = re.sub("Bearer\\s+", "", bearer_token)
+
+        # Attempt getting IT's OIDC configuration.
+        oidc_response = requests.get(url=self.oidc_configuration_url)
+        if not status.is_success(oidc_response.status_code):
+            logger.error(
+                f"Unable to get the OIDC configuration payload when attempting to validate a JWT token. Response code:"
+                f"{oidc_response.status_code}. Response body: {oidc_response.content}"
+            )
+            raise UnableMeetPrerequisitesError()
+
+        # Attempt getting their public certificates' URL.
+        jwks_uri = oidc_response.json()["jwks_uri"]
+        if not jwks_uri:
+            logger.error(
+                f"Unable to extract the JWKs' URI when attempting to validate a JWT token. Actual payload:"
+                f"{oidc_response.content}"
+            )
+            raise UnableMeetPrerequisitesError()
+
+        # Attempt getting their public certificates.
+        jwks_certificates_response = requests.get(url=jwks_uri)
```

Comment:
A lot of the code in this method appears to be getting the cert/key to verify the signature.  I suspect the cert/key doesn't change very often.  I wonder if it would be good to cache the key for a while (seconds...in case it does change?).  Just a thought.

Reply:
For comparison, we cache this public key in the gateway for 8 hours because of this as well.

Reply:
I have seen in the IT docs that they say that the keys should be downloaded at least daily, that's why I have set a 24h cache. Should I reduce to 8 as well?

Reply:
I've reduced the cache to 8 hours.